// react-tetris-overlay/src/components/SubmitForm.js
import React from 'react'
import styled from 'styled-components'
import { appConstants } from '../utilities'
import { Button } from './Button'

export class SubmitForm extends React.Component {
  state = { playerName: '' }
  render() {
    return (
      <Container>
        <Input
          placeholder='My name is...'
          onChange={this.handleInputChange}
          value={this.state.playerName}
        />
        <Button onClick={this.handleSubmit}>Submit score</Button>
      </Container>
    )
  }

  handleInputChange = e => {
    this.setState({ playerName: e.target.value })
  }

  handleSubmit = () => this.props.onSubmit(this.state.playerName)
}

const Input = styled.input`
  border: none;
  width: calc(100% - 10px);
  margin-left: 4px;
  padding: 2px;
  font-size: 1rem;
  color: ${appConstants.colors.black};

  ::placeholder {
    color: ${appConstants.colors.gray};
  }

  :focus {
    outline: none;
  }
`
const Container = styled.div`
  border: 2px solid ${appConstants.colors.black};
  position: absolute;
  top: -2px;
  left: 100%;
  width: 200px;
  padding-top: 4px;
  padding-bottom: 4px;
`
